### ISSUE TITLE
Incompatibility with Gpg4Win -> Decryption failure when using multiple recipients!

### ISSUE THREAD

**Issue by Draghtnod** [open-keychain/open-keychain#1016]
Using Gpg4Win 2.2.3 (based on GnuPG 2.0.26) and OpenKeyChain 3.1.2.

Enrcypt an ASCII armored text with Gpg4Win and enter more than one recipient. OpenKeyChain will fail to decrypt this with an integrity check error. Using only one recipient will work without errors.

**Comment by Valodim** [comment]
I just added a [unit test](https://github.com/open-keychain/open-keychain/commit/5e9d8ca8c6af6a3c274208799cd4b2a22de2d843) which tests encrypt+decrypt to multiple keys. This tests succeeds, so this doesn't seem to happen on just the "multiple recipients" condition. @Draghtnod Can you be more specific about when this error occurs? Or, even better, provide me with an example which produces the error when decrypting.

**Comment by Draghtnod** [comment]
Sorry for my second comment. I appreciate your time and efforts for this project.

I digged a bit deeper into this issue. This bug only occures if you encrypt with Gpg4Win and decrypt with OpenKeychain and only if you encrypt without signing and only with some public keys. I couldn't determine why this only happens with some keys. I was not able to reproduce the creation of such a malicious key, but i know that the one i use was initially created by OpenKeychain.
Gpg4Win to Gpg4Win works.
OpenKeychain to OpenKeychain works.
OpenKeychain to Gpg4Win works.
Gpg4Win to Openkeychain does not work.

Private Key1: http://pastebin.com/TqDYuWJr
Password: test
Encrypted for Key1 and another public (the bad) one with Gpg4Win: http://pastebin.com/KUg28xZy
This fails to decrypt on my OpenKeychain with an integrity check failure.

**Comment by Draghtnod** [comment]
The bad public key has two subkeys that are flagged for encryption. Maybe the Integrity check within OpenKeychain checks the flags of the subkeys and immediately fails if the first subkey with an encryption flag does not match the one that was used for encryption? Just a guess..

**Comment by Valodim** [comment]
Oh boy.

Ok so, thanks to your test case I was able to pinpoint the issue: for some reason, gpg4win creates a [symmetrically encrypted data packet (tag 9)](https://tools.ietf.org/html/rfc4880#page-45) instead of a [symmetrically encrypted integrity protected data packet (tag 18)](https://tools.ietf.org/html/rfc4880#page-49). 

I checked the key you provided, and it does specify that we support modification detection.
> 	Hashed Sub: features(sub 30)(critical)(1 bytes)
>		Flag - Modification detection (packets 18 and 19)

So what is happening here is that the packet is correctly encrypted, but OpenKeychain rejects it because it may have been tampered with. In the rfc, there is a bit on this in the [security considerations](https://tools.ietf.org/html/rfc4880#page-83).

I have another idea, just a second.

**Comment by Valodim** [comment]
(got it, will elaborate later)

**Comment by Valodim** [comment]
trying to keep it short: it's not gpg4win related, just gpg and OpenKeychain.

firstly, gpg is behaving conservatively in an arguably correct way by creating non-modification-protected encrypted data if *any* of the recipients doesn't have the "supports mdc" flag set. this is the more compatible but slightly less secure approach compared to always using mdc, which is what OpenKeychain does.

so why does it do that in this case? this one is actually our fault in a very non-obvious way. user ids are bound to a key using self certificates, and these certificates also contain information about preferred algorithms and supported capabilities including mdc. OpenKeychain correctly issues these when a key is created or modified. certifying a *foreign* key works by issuing the same type of certificate, but for another key and those certificates do not contain preferences for obvious reasons. the problem is that OpenKeychain allows creating this type of certificate by a key for itself, which overrides older self-certificates. if you look at the key you used for encryption above (luca) you will notice there is a cert made on Oct16, which is the correct one, and one made on Oct31 which has no preferences. You can fix that by performing any "Edit key" operation which affects the user ids (like changing expiry)

tl;dr: Allowing the user to use "Certify identities" of a key itself is not only pointless, but harmful, and we should make sure that doesn't happen

As a first step I made the error given for this case more descriptive (2192a6baddd09ae81f86a8bfd5e844086154f5ba)

**Comment by Valodim** [comment]
fixed this in the pgp routines (47ace7cea31ee794ed88bdf4163dd38fc33e8fc5) and added test cases (0c3f9ae0a67cb98b2ac3ac59985aec03e6a15861). it's not indicated in the ui, but a late error should be fine in this case I think.

thank you for your bug report.

**Comment by Draghtnod** [comment]
Great thank you! This should prevent this bug for the future and i know now how to fix the messed up key of my friend. :)

**Comment by isermann** [comment]
Hey, I get this integrity check error since version 3.2.0-3.2.3. Version 3.1 was working flawlessly, though. Receiving emails in K9 encrypted with Gpg4Win to just one recipient.
Can you please look into it again?

And is there a way to get a pre-compiled version 3.1.2, as F-Droid only hosts the three latest versions?

**Comment by dschuermann** [comment]
@isermann Please open a new issue
